From 5b2e87901b0220f8c32f3eb0de16a61de5be889f Mon Sep 17 00:00:00 2001
From: Michael Olbrich <m.olbrich@pengutronix.de>
Date: Wed, 23 Jan 2013 14:12:16 +0100
Subject: [PATCH] service: make sure the watchdog timer is not restarted while
 stopping

A watchdog notification may be handled after the watchdog timer was stopped
while stopping the service. As a result the timer is restarted and the
service may be restarted as well.
The watchdog timestamp is initially set during startup in
service_enter_start_post() and cleared when the timer is stopped. Therefore
it can be used as an indication if the timer should be reset.
(cherry picked from commit 90527fbb2c48ffda5c6d8f232f8993a90b2632a4)
---
 src/core/service.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/core/service.c b/src/core/service.c
index 8eefd2d..0218e76 100644
--- a/src/core/service.c
+++ b/src/core/service.c
@@ -3415,7 +3415,8 @@ static void service_notify_message(Unit *u, pid_t pid, char **tags) {
         if (strv_find(tags, "WATCHDOG=1")) {
                 log_debug_unit(u->id,
                                "%s: got WATCHDOG=1", u->id);
-                service_reset_watchdog(s);
+                if (dual_timestamp_is_set(&s->watchdog_timestamp))
+                        service_reset_watchdog(s);
         }
 
         /* Notify clients about changed status or main pid */
